### ISSUE TITLE
Execution.zip can result in recomputation

### ISSUE THREAD

**Issue by johnynek** [twitter/scalding#1237]
Consider the case with #1217 and we do
```scala
val a = in.make(storage)
val a1 = a.flatMap(fn1)
val a2 = a1.flatMap(fn2)
val a3 = a1.flatMap(fn3)
val a4 = a2.zip(a3)
```

here is a classic diamond. I think your case can be relabeled to look like this. a = Execution.getMode. So, why is it evaluating everything twice. I think the answer is a subtle distinction between the evaluation of Execution[T] => Future[T] and the evaluations of cascading flows => T. Really what happens in some cases is `Execution[T] => Flow`, then `Flow => Future[T]`. I think the problem here is that we are sharing an evaluation of `Execution[T] => Future[T]` to on the first step, but then you have a `Future[TypedPipe[U]]`, which is actually still a program that will be run to create some sequence of `U`. So, both sides of the diamond share the same `TypedPipe[U]`, but that does not mean they will share the same on disk (unfortunately).

This case is a bit confusing, and I think it brings up that we need to:

1) be a little more clear about what the contracts and recomputation scenarios are in Execution. Right now, I guess the truth is when you use zip, you could wind up recomputing everything twice (although flatMap should not have that problem, but could reduce parallelism and increase latency).
2) Try to minimize these cases more so that there are fewer cases (hopefully none) of recomputation. I'm not sure that is always possible.

This may be a particular issue with the `writeThrough` method. It would be good to have some tests here. We do have one test:

https://github.com/twitter/scalding/blob/develop/scalding-core/src/test/scala/com/twitter/scalding/typed/ExecutionTest.scala#L184

but it actually dodges the issue. Yes we render 1 typed-pipe, but putting that in two flows with zip is going to cause it to be created twice.

**Comment by johnynek** [comment]
@ianoc suggests something leading to the following:

Should we be able to always ensure the same computation from:
```scala
val first = a.flatMap(f).zip(a.flatMap(g))
val second = a.flatMap { r => f(r).zip(g(r)) }
```
Right now, this is not the case. .zip on a flatMapped Execution is different from a .zip on pure cascading operation (from .writeExecution for instance). In the first case, we hide from cascading the structure, but in the second case, we may implement zip by merging flows in cascading, which should give a better plan.

**Comment by johnynek** [comment]
should be closed by #1253
